File: assignments/week-12/0053-maximum-subarray/solution.java
class Solution {
    public int maxSubArray(int[] nums) {
        int maxSubArrSum = Integer.MIN_VALUE;
        int currSubArrSum = 0;

        for(int num : nums){
            currSubArrSum = Math.max(num, currSubArrSum + num);

            maxSubArrSum = Math.max(maxSubArrSum, currSubArrSum);
        }

        return maxSubArrSum;
    }
}

File: assignments/week-12/0053-maximum-subarray/solution.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        memoDP = [0 for _ in range(len(nums))]
        max_sub_arr_sum = nums[0]

        memoDP[0] = nums[0]

        for idx_i in range(1, len(nums)):
            curr_num = nums[idx_i]
            memoDP[idx_i] = max(memoDP[idx_i - 1] + curr_num, curr_num)

            max_sub_arr_sum = max(max_sub_arr_sum, memoDP[idx_i])
        
        return max_sub_arr_sum